docker-images/syntax-highlighter/crates/scip-treesitter-cli/src/main.rs

```diff
@@ -0,0 +1,326 @@
+        filenames: Vec<String>,
+
+        /// Analysis mode
+        #[arg(short, long, default_value = "full")]
+        mode: AnalysisMode,
+
+        /// Fail on first error
+        #[arg(long, default_value_t = false)]
+        strict: bool,
+
+        /// Project root to write to SCIP index
+        #[arg(short, long, default_value = "./")]
+        project_root: String,
+    },
+}
+
+struct Options {
+    analysis_mode: AnalysisMode,
+    /// When true, fail on first encountered error
+    /// Otherwise errors are logged but they don't
+    /// interrupt the process
+    strict: bool,
+}
+
+pub fn main() {
+    let cli = Cli::parse();
+
+    match cli.command {
+        Commands::Index {
+            language,
+            out,
+            filenames,
+            mode,
+            strict,
+            project_root,
+        } => index_command(
+            language,
+            filenames,
+            PathBuf::from(out),
+            PathBuf::from(project_root),
+            Options {
+                analysis_mode: mode,
+                strict,
+            },
+        ),
+    }
+}
+
+fn index_command(
+    language: String,
+    filenames: Vec<String>,
+    out: PathBuf,
+    project_root: PathBuf,
+    options: Options,
+) {
+    let p = BundledParser::get_parser(&language).unwrap();
+
+    let mut index = scip::types::Index {
+        metadata: Some(scip::types::Metadata {
+            tool_info: Some(scip::types::ToolInfo {
+                name: "scip-treesitter-cli".to_string(),
+                version: clap::crate_version!().to_string(),
+                arguments: vec![],
+                ..Default::default()
+            })
+            .into(),
+            project_root: format!(
+                "file://{}",
+                project_root
+                    .canonicalize()
+                    .expect("Failed to canonicalize project root")
+                    .display()
+            ),
+            ..Default::default()
+        })
+        .into(),
+        ..Default::default()
+    };
+
+    let bar = ProgressBar::new(filenames.len() as u64);
+
+    bar.set_style(
+        ProgressStyle::with_template(
+            "[{elapsed_precise}] {bar:40.cyan/blue} {pos:>7}/{len:7}\n {msg}",
+        )
+        .unwrap()
+        .progress_chars("##-"),
+    );
+
+    for (_, filename) in filenames.iter().enumerate() {
+        let contents = std::fs::read(filename).unwrap();
+        bar.set_message(filename.clone());
+        bar.inc(1);
+        match index_content(contents, p, &options) {
+            Ok(mut document) => {
+                document.relative_path = filename.to_string();
+                index.documents.push(document);
+            }
+            other => {
+                if options.strict {
+                    other.unwrap();
+                } else {
+                    eprintln!("Failed to extract locals from {filename}: {:?}", other)
+                }
+            }
```

Comment:
Stylistic suggestion:

```suggestion
            Err(error) => {
                if options.fail_fast {
                    panic!("Failed to index {filename}: {:?}", error);
                } else {
                    eprintln!("Failed to index {filename}: {:?}", error);
                }
            }
```

1. Extracting the error avoids the `unwrap` (which will always panic, so it's a bit confusing to read).
2. The error message about "Failed to extract locals" is technically not necessarily correct since we're also extracting global symbols, right?